src/components/Form/Widgets/FileWidget.tsx

```diff
@@ -1,248 +1,478 @@
+	maxSize: number,
+) => {
+	switch (code) {
+		case 'file-too-large':
+			return `File is larger than ${fileSizeToReadable(maxSize)}`;
+		default:
+			return message;
+	}
+};
 
-export interface OnFileReadSuccessParams {
+export interface OnFileReadParams {
 	dataUrl: string;
 	uploadedFile: File;
 }
 
-export interface OnFileReadErrorParams {
-	uploadErrors: string[];
-}
-
-// We need to forward uploadErrors onChange because RJSF does not support widget modifying
-// the default error handler.See: https://github.com/rjsf-team/react-jsonschema-form/issues/2718
-export type OnFileReadParams = OnFileReadSuccessParams | OnFileReadErrorParams;
-
 type FileWidgetProps = WidgetProps & {
+	// TODO: Currently it's not officially supported to raise errors via a custom widget's onChange, any attempt to do so would be a workaround
+	// See: https://github.com/rjsf-team/react-jsonschema-form/issues/2718
+	// There is a PR open to address this, so we should keep an eye on it and revisit this widget's "error-raising" when it's merged
+	// See: https://github.com/rjsf-team/react-jsonschema-form/pull/4188
 	onChange: (params?: OnFileReadParams) => void;
 	value?: OnFileReadParams | string;
 };
 
 export const FileWidget = ({
 	onChange,
-	value,
 	multiple,
-	...props
+	schema,
+	uiSchema,
 }: FileWidgetProps) => {
-	const { schema, uiSchema } = props;
+	const [files, setFiles] = useState<AcceptedFile[]>([]);
+	const [errorFiles, setErrorFiles] = useState<FileRejection[]>([]);
+	const theme = useTheme();
+	const mobile = useMediaQuery(theme.breakpoints.down('md'));
 
-	const [loadingPercentage, setLoadingPercentage] = useState<
-		number | undefined
-	>(undefined);
-	const [errors, setErrors] = useState<string[]>([]);
+	const fileUploadId = useMemo(() => `file-upload-${crypto.randomUUID()}`, []);
 
 	const accept = uiSchema?.['ui:options']?.accept as Accept | undefined;
 	const maxSize = uiSchema?.['ui:options']?.maxSize as number | undefined;
-
-	const dataUrl = (
-		dataUrlValue: OnFileReadParams | string,
-	): string | undefined => {
-		if (typeof dataUrlValue === 'string') {
-			return dataUrlValue;
-		}
-
-		if ('dataUrl' in dataUrlValue) {
-			return dataUrlValue.dataUrl;
-		}
-	};
-
-	const isImage = (value: OnFileReadParams | string) => {
-		return dataUrl(value)?.startsWith('data:image/');
-	};
+	const disabled = uiSchema?.['ui:options']?.disabled;
+	const disabledReason = uiSchema?.['ui:options']?.disabledReason as
+		| string
+		| undefined;
+	const browseResourceName = uiSchema?.['ui:options']?.browseResourceName as
+		| string
+		| undefined;
+	const isSmall =
+		(uiSchema?.['ui:options']?.variant as 'large' | 'small' | undefined) ===
+		'small';
+	const uploadIcon = uiSchema?.['ui:options']?.uploadIcon as
+		| IconDefinition
+		| undefined;
+	const descriptiveText = uiSchema?.['ui:options']?.descriptiveText as
+		| string
+		| undefined;
+	const hideUploadedFiles = uiSchema?.['ui:options']?.hideUploadedFiles as
+		| boolean
+		| undefined;
```

Comment:
what about something like this to avoid all these type casts ? (feel free to update it, just a basic idea)

```
type UIOptions = {
	accept?: Accept;
	maxSize?: number;
	disabled?: boolean;
	disabledReason?: string;
	browseResourceName?: string;
	variant?: 'large' | 'small';
	uploadIcon?: IconDefinition;
	descriptiveText?: string;
	hideUploadedFiles?: boolean;
};

const options = uiSchema?.['ui:options'] as UIOptions | undefined;

const accept = options?.accept;
const maxSize = options?.maxSize;
const disabled = options?.disabled;
const disabledReason = options?.disabledReason;
const browseResourceName = options?.browseResourceName;
const isSmall = options?.variant === 'small';
const uploadIcon = options?.uploadIcon;
const descriptiveText = options?.descriptiveText;
const hideUploadedFiles = options?.hideUploadedFiles;
```

Reply:
I was thinking about the same, don't remember why I dropped the idea, let me do it